db/system_keyspace.cc

```diff
@@ -2676,8 +2676,13 @@ service::topology_features system_keyspace::decode_topology_features_state(::sha
     for (auto& row : *rs) {
         raft::server_id host_id{row.get_as<utils::UUID>("host_id")};
         service::node_state nstate = service::node_state_from_string(row.get_as<sstring>("node_state"));
-        if (row.has("supported_features") && nstate == service::node_state::normal) {
-            ret.normal_supported_features.emplace(host_id, decode_features(deserialize_set_column(*topology(), row, "supported_features")));
+        if (row.has("supported_features")) {
+            auto features = decode_features(deserialize_set_column(*topology(), row, "supported_features"));
+            if (nstate == service::node_state::normal) {
```

Comment:
Hmm, right. I think it implicitly works correctly right now, because adding nodes and transitioning nodes has more priority in the topology coordinator right now than enabling features.

There are two ways in which a node can be protected from a hazardous situation (enabling a feature when a node in the cluster does not support it):

1. Include that node when calculating the intersection of supported features of all nodes,
2. Forbid enabling features when the node is in some state

Currently, we apply 1. for normal nodes, and 2. for others (bootstrapping nodes, just-joining nodes).

I suppose we could get rid of 2. completely if we:

- Include transitioning (i.e. bootstrapping) nodes when calculating features to enable and send the read barrier request to them as well,
- Do a topology state machine transition before sending join_node_response, like I mentioned here: https://github.com/scylladb/scylladb/pull/15196#discussion_r1309819542

Reply:
Hm, TBH I don't see advantages of one solution or the other.
@gleb-cloudius WDYT?

But if we stay with the current approach -- i.e. the precondition of enabling is that all nodes are NORMAL -- this place should have a comment to explain why the code is safe (we don't need to interset with non-normal nodes -- they must be empty).

Reply:
I think we should be able to enable features while some topology operations are still waiting to be executed in the request queue. It is fine to wait for running operation (bootstrap/replace) to complete though.

Reply:
Now it's safe to enable features when there are nodes in `none` state, but not when there are transitioning nodes. Adjusted the check in `enable_features` and clarified explanations in the comments why it works and considering only normal nodes for barrier and feature calculation is safe.